### ISSUE TITLE
Upgrade fails because the URL for the Saltstack GPG key no longer exists

### ISSUE THREAD

**Issue by RobotsOnDrugs** [REMnux/remnux-cli#86]
https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3001 does not exist anymore, so the installer cannot download the GPG key in order to verify the SaltStack installation.

![image](https://user-images.githubusercontent.com/306289/176793241-c34b3ca3-8a81-4dac-b504-10d0f2ff8e5e.png)

**Comment by lennyzeltser** [comment]
@RobotsOnDrugs, please follow these steps to install the latest version of the REMnux updater, which should address the issue:

Download the installer:

```
wget https://REMnux.org/remnux-cli
```

Validate that the SHA-256 hash of the downloaded file to make sure it matches this expected value:
`23c7f4eefa7599ea2c4156f083906ea5fd99df18f306e4bb43eec0430073985a`

To generate the hash of your  file, run:
```
sha256sum remnux-cli
```
Set up the REMnux installer:
```
mv remnux-cli remnux
chmod +x remnux
sudo mv remnux /usr/local/bin
```

**Comment by ctfmastery** [comment]
Hello @lennyzeltser, the updated file doesn't fully fix it. 

Altought it begins update/upgrade, it still refers the nonexistent URL and ends up prematurely after upgrading most softwares. 

```
> remnux-cli@1.3.4.2.g87c65ef
> remnux-version: v2021.15.2

NOTICE: Fixing incorrect Saltstack version configuration.
Installing and configuring Saltstack properly ...
> downloading v2022.26.3
>> downloading remnux-salt-states-v2022.26.3.tar.gz.asc
>> downloading remnux-salt-states-v2022.26.3.tar.gz.sha256
>> downloading remnux-salt-states-v2022.26.3.tar.gz.sha256.asc
>> downloading remnux-salt-states-v2022.26.3.tar.gz
> validating file remnux-salt-states-v2022.26.3.tar.gz
> validating signature for remnux-salt-states-v2022.26.3.tar.gz.sha256
> extracting update remnux-salt-states-v2022.26.3.tar.gz
> using previous mode: dedicated
> upgrading/updating to v2022.26.3
>> Log file: /var/cache/remnux/cli/v2022.26.3/saltstack.log

>> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3001 focal main
>> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3004 focal main
>> Running: deb [arch=amd64] https://download.docker.com/linux/ubuntu focal stable
>> Running: deb http://download.draios.com/stable/deb stable-amd64/
>> Running: deb http://www.inetsim.org/debian/ binary/
```
....
```
>> Running: remnux-theme
>> Running: ssh
Update returned exit code not zero
Error: Update returned exit code not zero
    at ChildProcess.<anonymous> (/snapshot/remnux-cli/remnux-cli.js:569:23)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)


Sometimes problems occur due to network or server issues when
downloading packages, in which case retrying your operation
a bit later might lead to good results.

To determine the nature of the issue, please review the
saltstack.log file under /var/cache/remnux/cli in the
subdirectory that matches the REMnux version you're installing.
Search for the log file for "result: false" messages and look at
the surrounding lines to diagnose the issue.

For assistance go to https://github.com/REMnux/remnux-cli/issues
```

Also, even after the upgrade, the update function downloads remnux-cli older version back again:
```
remnux@remnux:~$ remnux update
> remnux-cli@1.3.4.2.g87c65ef
> remnux-version: v2021.15.2

> downloading v2021.15.2
>> downloading remnux-salt-states-v2021.15.2.tar.gz.asc
>> downloading remnux-salt-states-v2021.15.2.tar.gz.sha256
>> downloading remnux-salt-states-v2021.15.2.tar.gz.sha256.asc
>> downloading remnux-salt-states-v2021.15.2.tar.gz
> validating file remnux-salt-states-v2021.15.2.tar.gz
> validating signature for remnux-salt-states-v2021.15.2.tar.gz.sha256
> extracting update remnux-salt-states-v2021.15.2.tar.gz
> using previous mode: dedicated
> upgrading/updating to v2021.15.2
>> Log file: /var/cache/remnux/cli/v2021.15.2/saltstack.log
```

**Comment by ctfmastery** [comment]
after a few attempts of upgrade + update + upgrade, it actually completed the upgrage:


```
...
>> Running: ssh
>> Running: remnux-dedicated

>> COMPLETED SUCCESSFULLY! Success: 815, Failure: 0


>> Please reboot to make sure all settings go into effect.
```

And then after the reboot the system crashs a few times and has to be forcedly rebooted before properly initialize.

**Comment by RobotsOnDrugs** [comment]
Running `remnux upgrade` after the manual update did resolve the issue, but FYI to anyone who runs into this issue, running `remnux update` before the upgrade will clobber remnux-cli and install the old version.

I would say this bug is not that it can't retrieve the PGP key, but rather that the installer in the appliance seems to lack the ability to upgrade itself. I would say that this workaround should be in the documentation, at least until the appliance is updated.

**Comment by rtkcgrantcharov** [comment]
> Hello @lennyzeltser, the updated file doesn't fully fix it.
> 
> Altought it begins update/upgrade, it still refers the nonexistent URL and ends up prematurely after upgrading most softwares.
> 
> ```
> > remnux-cli@1.3.4.2.g87c65ef
> > remnux-version: v2021.15.2
> 
> NOTICE: Fixing incorrect Saltstack version configuration.
> Installing and configuring Saltstack properly ...
> > downloading v2022.26.3
> >> downloading remnux-salt-states-v2022.26.3.tar.gz.asc
> >> downloading remnux-salt-states-v2022.26.3.tar.gz.sha256
> >> downloading remnux-salt-states-v2022.26.3.tar.gz.sha256.asc
> >> downloading remnux-salt-states-v2022.26.3.tar.gz
> > validating file remnux-salt-states-v2022.26.3.tar.gz
> > validating signature for remnux-salt-states-v2022.26.3.tar.gz.sha256
> > extracting update remnux-salt-states-v2022.26.3.tar.gz
> > using previous mode: dedicated
> > upgrading/updating to v2022.26.3
> >> Log file: /var/cache/remnux/cli/v2022.26.3/saltstack.log
> 
> >> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3001 focal main
> >> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3004 focal main
> >> Running: deb [arch=amd64] https://download.docker.com/linux/ubuntu focal stable
> >> Running: deb http://download.draios.com/stable/deb stable-amd64/
> >> Running: deb http://www.inetsim.org/debian/ binary/
> ```
> 
> ....
> 
> ```
> >> Running: remnux-theme
> >> Running: ssh
> Update returned exit code not zero
> Error: Update returned exit code not zero
>     at ChildProcess.<anonymous> (/snapshot/remnux-cli/remnux-cli.js:569:23)
>     at ChildProcess.emit (events.js:315:20)
>     at maybeClose (internal/child_process.js:1021:16)
>     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
> 
> 
> Sometimes problems occur due to network or server issues when
> downloading packages, in which case retrying your operation
> a bit later might lead to good results.
> 
> To determine the nature of the issue, please review the
> saltstack.log file under /var/cache/remnux/cli in the
> subdirectory that matches the REMnux version you're installing.
> Search for the log file for "result: false" messages and look at
> the surrounding lines to diagnose the issue.
> 
> For assistance go to https://github.com/REMnux/remnux-cli/issues
> ```
> 
> Also, even after the upgrade, the update function downloads remnux-cli older version back again:
> 
> ```
> remnux@remnux:~$ remnux update
> > remnux-cli@1.3.4.2.g87c65ef
> > remnux-version: v2021.15.2
> 
> > downloading v2021.15.2
> >> downloading remnux-salt-states-v2021.15.2.tar.gz.asc
> >> downloading remnux-salt-states-v2021.15.2.tar.gz.sha256
> >> downloading remnux-salt-states-v2021.15.2.tar.gz.sha256.asc
> >> downloading remnux-salt-states-v2021.15.2.tar.gz
> > validating file remnux-salt-states-v2021.15.2.tar.gz
> > validating signature for remnux-salt-states-v2021.15.2.tar.gz.sha256
> > extracting update remnux-salt-states-v2021.15.2.tar.gz
> > using previous mode: dedicated
> > upgrading/updating to v2021.15.2
> >> Log file: /var/cache/remnux/cli/v2021.15.2/saltstack.log
> ```

Pretty sure this error is caused by #88 which I just logged. I saw: 
```
>> Running: remnux-theme
>> Running: ssh
Update returned exit code not zero
Error: Update returned exit code not zero
    at ChildProcess.<anonymous> (/snapshot/remnux-cli/remnux-cli.js:569:23)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
```
occur when I installed from scratch and when I attempted to upgrade in `dedicated` mode (once I got past the missing repo and PGP key errors).

**Comment by digitalsleuth** [comment]
Hi @rtkcgrantcharov Can you provide your saltstack.log file for this? The references you saw to Saltstack above:
```bash
>> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3001 focal main
>> Running: deb [arch=amd64] https://repo.saltproject.io/py3/ubuntu/20.04/amd64/3004 focal main
```
Is the process of removing the 3001 focal, and adding the 3004 focal. It's not actually installing 3001, it's removing it in favor of 3004.

Your saltstack.log file will contain the actual error you received during this install which, I'm guessing by #88 is the original cause. However, it's best to actually review the entire log first.

**Comment by ctfmastery** [comment]
Hi @digitalsleuth thank you for your reply.

Sure, here you are the logs:

[saltstack (old).log](https://github.com/REMnux/remnux-cli/files/9087656/saltstack.old.log)
[saltstack (new).log](https://github.com/REMnux/remnux-cli/files/9087658/saltstack.new.log)

I got it working already, but I think it contains a bug, since I had to run it multiple times to get it to download the correct version and get working properly. And I don't know if it appended the logs or replaced them afterall.
